### PR TITLE
Fix `ScalarValue` handling of NULL values for ListArray

### LINKED ISSUE #7968: Incorrect and inconsistency handling of NULL values in `ScalarValue` for ListArray
### Describe the bug

* `try_from_array` produces inconsistent `ScalarValue::List` for valid and non-valid (NULL) values

    For NULL value in ListArray, `try_from_array` produces `ScalarValue::List(new_null_array(&DataType::Null, 0))` which isn't consistent with the case when invoking `try_from_array` on valid value of `ListArray`.

    Fro valid value, `try_from_array` produces a `ListArray` containing single element of the specified value in provided `ListArray`. For example, if input array is `ListArray[[1, 2, 3], [4, 5, 6]]]`, `try_from_array(arr, 0)` will produce `ListArray[[1, 2, 3]]`.

    For NULL value, `try_from_array` currently produces a null array of`DataType::Null` type. Not only it is not a `ListArray`, the data type is not the element type of provided `ListArray` too.

* `ScalarValue::List` has same value and datatype for `ListArray` and nested `ListArray`'s NULL values
    NULL values from `ListArray` and nested `ListArray` are the same `ScalarValue::List(new_null_array(&DataType::Null, 0))`. So their datatypes are the same. This is also incorrect.

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #7968.

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -3885,6 +3889,77 @@ mod tests {
...
+
+        let nested_list_data_type = DataType::List(Arc::new(Field::new(
+            "item",
+            DataType::List(Arc::new(Field::new("item", DataType::Int32, true))),
+            true,
+        )));
+        let data_type = &nested_list_data_type;
+        let nested_list_scalar: ScalarValue = data_type.try_into().unwrap();
+
+        assert_ne!(list_scalar, nested_list_scalar);
```
Comment (viirya) [inline datafusion/common/src/scalar.rs:3960]: Previously the `ScalarValue`s for list and nested list NULL values are equal, that is incorrect.

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -2234,28 +2239,20 @@ impl ScalarValue {
             }
             DataType::Utf8 => typed_cast!(array, index, StringArray, Utf8),
             DataType::LargeUtf8 => typed_cast!(array, index, LargeStringArray, LargeUtf8),
-            DataType::List(nested_type) => {
+            DataType::List(_) => {
                 let list_array = as_list_array(array);
-                let arr = match list_array.is_null(index) {
-                    true => new_null_array(nested_type.data_type(), 0),
```
Comment (viirya) [inline datafusion/common/src/scalar.rs:2240]: This condition is redundant. `try_from_array` handles NULL value at the beginning.

File: datafusion/common/src/scalar.rs [hunk 3]
```diff
@@ -2234,28 +2239,20 @@ impl ScalarValue {
...
-                    }
-                };
+                let nested_array = list_array.value(index);
+                // Produces a single element `ListArray` with the value at `index`.
+                let arr = Arc::new(wrap_into_list_array(nested_array));
 
                 ScalarValue::List(arr)
             }
             // TODO: There is no test for FixedSizeList now, add it later
-            DataType::FixedSizeList(nested_type, _len) => {
```
Comment (viirya) [inline datafusion/common/src/scalar.rs:2250]: We can add some tests for `FixedSizeList`. I can do it in other follow up.

File: datafusion/common/src/scalar.rs [hunk 4]
```diff
@@ -3885,6 +3889,77 @@ mod tests {
...
     }
 
+    #[test]
+    fn scalar_try_from_array_list_array_null() {
+        let list = ListArray::from_iter_primitive::<Int32Type, _, _>(vec![
+            Some(vec![Some(1), Some(2)]),
+            None,
+        ]);
+
+        let non_null_list_scalar = ScalarValue::try_from_array(&list, 0).unwrap();
```
Comment (viirya) [inline datafusion/common/src/scalar.rs:3899]: Previously, for a ScalarValue NULL value from ListArray, its datatype is `DataType::Null`. That's is not correct.

File: datafusion/common/src/scalar.rs [hunk 5]
```diff
@@ -1699,22 +1702,24 @@ impl ScalarValue {
...
                     valid.append(false);
-                } else {
-                    let list_arr = as_list_array(&arr);
-                    let arr = list_arr.values().to_owned();
-                    offsets.push(arr.len());
-                    elements.push(arr);
-
-                    // Element is valid
-                    valid.append(true);
+                    continue;
```
Comment (jayzhan211) [inline datafusion/common/src/scalar.rs:1714]: I think if-else statement here is more readable.

**Review by jayzhan211 — APPROVED** [review]
LGTM

**Review by comphead — APPROVED** [review]
lgtm
thanks @viirya

**Review by jackwener — APPROVED** [review]
Good catch, thanks @viirya

**Comment by viirya** [comment]
Thank you @alamb @comphead @jackwener @jayzhan211

**Comment by alamb** [comment]
> Thank you @alamb @comphead @jackwener @jayzhan211

Team effort 🦾

**Pull request merged**